### ISSUE TITLE
[no-unnecessary-condition] What's the point of `ignoreRhs: false` option?

### ISSUE THREAD

**Issue by ark120202** [typescript-eslint/typescript-eslint#1017]
I am probably missing something, but reading through all of comments in original PR (#699) I couldn't find a reason to include `ignoreRhs` option. It probably made sense for `strict-boolean-expressions`, but I can't think of any cases where I'd want `return items.length && items[0].toUpperCase()` to be reported, since there is no _condition_ part.

At the same time I think RHS should always be reported when used in a condition, like:
```ts
function test(foo: { bar: true } | undefined) {
  if (foo && foo.bar) {}
  //         ^^^^^^^
}
```

(note: `items.length && items[0].toUpperCase()` from docs isn't correct example for this rule, since `toUpperCase` returns possibly falsy value)

EDIT: Actually, I think it might make sense for this to be behavior for `strict-boolean-expressions` as well. Looks like that option was added just to avoid breaking change, see https://github.com/palantir/tslint/issues/4447.

**Comment by sindresorhus** [comment]
Here's a use-case: https://github.com/xojs/eslint-config-xo-typescript/issues/17

// @EdJoPaTo @yangmingshan

**Comment by bradzacher** [comment]
With rules it's much better to implement them with zero options and relax them over time as people find cases that are reasonable to want to ignore.

In this case, the rule was implemented strictly, making the assumption that you didn't use short circuiting - `&&` is a boolean operator, after all.

Then someone decided to add the option, because they felt that it should be safe to allow short circuiting.

Whilst adding a new option is a minor feature bump, turning on the option as default is a major breaking bump.

I'm not sure if it's a good idea to turn it on by default - in 3.7 they are implementing optional chaining and nullish coalescing, which will handle many of the cases people use short circuiting for.

That's a point for discussion however

**Comment by ark120202** [comment]
> In this case, the rule was implemented strictly, making the assumption that you didn't use short circuiting

I guess that's about `strict-boolean-expressions` rule? In `no-unnecessary-condition` it was there since the first commit: https://github.com/typescript-eslint/typescript-eslint/pull/699/commits/5ec23e1f6fe98d689e32541854343d8e86b10c2a#diff-1dcdebcfacc649bba21e6ffe12f28ba1R43.

> && is a boolean operator, after all

That's a good point, I can see how someone would like to enforce that and it fits `strict-boolean-expressions` well, but I don't see why it should be in `no-unnecessary-condition`

**Comment by bradzacher** [comment]
Sorry, I forget which one had what at PR time.
In that case it was that the PR author created the rule via copy pasting strict-boolean-expressions, and then didn't think about changing the option to be default on.
I think I had a busy week, so I didn't review it in depth enough to make a decision on if the options should be on or if it should even exist.

I'm not sure if there is a case for having the option turned off in no-unnecessary-condition.

**Comment by Retsam** [comment]
Original author here, unless I'm misreading @ark120202's comment, it's not a question of changing the default; it's that the option is unnecessary and can be removed altogether.  

And I think they're right.  Instead of using an option to determine when the right-hand-side should be checked, it should be done by the context of the expression.  

In the context of a condition, (e.g. `if(foo && bar)`) it should always check `bar` to ensure the condition is necessary.  In other contexts (e.g. `return foo && bar;` it doesn't ever make sense to check `bar`.

I'll see if I can make the changes to make this happen.  I'm not sure if it'll count as a breaking change to remove an option and unify both cases.

**Comment by bradzacher** [comment]
Related: #1118

Removing an option is a breaking change. It is not backward compatible.

**Comment by jzarzeckis** [comment]
I'm quite confused with the cases when you have values that are potentially `undefined` on the left side.

Is this really the expected behaviour?
![image](https://user-images.githubusercontent.com/919350/70230272-a45bcb80-1760-11ea-970d-8de72aad262c.png)

I do have strict mode and `ignoreRhs` enabled, and foo is not always truthy.

Have I misconfigured something, or am I really expected to refactor such snippets to:
```ts
const bar: string[] = foo === undefined ? ['hello'] : foo
```

This seems less convenient.

Or is this expected to be fixed in #1163 ?

**Comment by glen-84** [comment]
@jzarzeckis Can you show the code?

Is it just this?

```ts
const foo: string[] | undefined = undefined;

const bar: string[] = foo || ['hello'];
```

... or is some code missing?

**Comment by jzarzeckis** [comment]
Just this. If that helps, I can also share my `eslintrc.json`.

@glen-84 - are you saying I'm alone with such behaviour?

**Comment by glen-84** [comment]
1. You should be seeing `value is always falsy`, not `value is always truthy`.
2. As far as I know, this is the correct behaviour, as the variable `foo` is last assigned the value `undefined`, and therefore the test will always fail (return `false` and use the default value of `['hello']`).

**Comment by jzarzeckis** [comment]
No no. The **type** of `foo` is `string[] | undefined`. And imho `undefined` is not `truthy`.

If strict null checking mode was disabled in tsconfig, I'd understand this behaviour, as it'd infer the type of foo as just `string[]`.

But since in my project I have `strict` mode enabled, I'd expect not to see such an eslint warning.

**Comment by ark120202** [comment]
@jzarzeckis the type of `foo` is `undefined`, because TypeScript is able to infer it from control flow. If you'd initialize `foo` with, for example `Math.random() > 0.5 ? undefined : ['a']`, eslint wouldn't show a warning there.

**Comment by glen-84** [comment]
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#control-flow-based-type-analysis

**Comment by jzarzeckis** [comment]
Okay - then how about this:
```ts
function fun(foo: string[] | undefined): string[] {
  const bar: string[] = foo || ['hello']

  return bar
}
```

![image](https://user-images.githubusercontent.com/919350/70247606-4b049400-1782-11ea-8e24-955747b1603f.png)

**Comment by glen-84** [comment]
I don't see any errors from that code. Are your dependencies up-to-date?

**Comment by jzarzeckis** [comment]
Hmm. I just upgraded to `2.10.0` and I still get the same warning. Registered #1308 , but if no-one else is able to reproduce - then I'm wondering what could I have configured incorrectly.

**Comment by bradzacher** [comment]
Commenting on an issue like this is not the place to report or discuss bugs.
please open a new issue and fill out the template.
